pkg/tempdir/tempdir.go

```diff
@@ -0,0 +1,306 @@
+}
+
+// initInstanceTempDir creates and locks a new unique temporary directory for this TempDir instance.
+// It acquires the tempDirGlobalLock during its operation to ensure atomicity
+// with other temp dir management operations.
+// The caller should ensure .Cleanup() is called eventually.
+func (td *TempDir) initInstanceTempDir() error {
+	if td.tempDirPath != "" && td.tempDirLock != nil {
+		logrus.Debugf("Temp dir %s already initialized", td.tempDirPath)
+		return nil
+	}
+	randomSuffix := stringid.TruncateID(stringid.GenerateRandomID())
+	tempDirName := TempDirPrefix + randomSuffix
+
+	tempDirLockPath := filepath.Join(td.rootDir, TempdirLockPrefix+tempDirName)
+	actualTempDirPath := filepath.Join(td.rootDir, tempDirName)
+
+	tempDirLock, err := lockfile.GetLockFile(tempDirLockPath)
+	if err != nil {
+		return fmt.Errorf("getting instance lock file %s for new temp dir %s: %w", tempDirLockPath, actualTempDirPath, err)
+	}
+
+	if err := tempDirLock.TryLock(); err != nil {
+		if errRm := os.RemoveAll(actualTempDirPath); errRm != nil && !os.IsNotExist(errRm) {
+			err = errors.Join(err, errRm)
+		}
+		return fmt.Errorf("locking instance lock for new temp dir %s failed: %w", actualTempDirPath, err)
+	}
+
+	if err := os.MkdirAll(actualTempDirPath, 0o700); err != nil {
+		return fmt.Errorf("creating temp directory %s failed: %w", actualTempDirPath, err)
+	}
+
+	td.tempDirPath = actualTempDirPath
+	td.tempDirLock = tempDirLock
+	td.tempDirLockPath = tempDirLockPath
+	return nil
+}
+
+// NewTempDir prepares a TempDir manager for a temporary directory in the specified rootDir.
+// The rootDir itself will be created if it doesn't exist.
+// It performs recovery of stale temporary directories within rootDir (removing those not in active use).
+// The actual instance-specific temporary directory is created and locked on the first call to Add().
+// Stale directory recovery (in NewTempDir) and instance directory creation (in Add via initInstanceTempDir).
+func NewTempDir(rootDir string) (*TempDir, error) {
+	if err := os.MkdirAll(rootDir, 0o700); err != nil {
+		return nil, fmt.Errorf("creating root temp directory %s failed: %w", rootDir, err)
+	}
+
+	td := &TempDir{
+		rootDir: rootDir,
+	}
+
+	if err := td.RecoverStaleDirs(); err != nil {
+		logrus.Warnf("Error during stale temp dir recovery in %s: %v", rootDir, err)
+	}
+
+	return td, nil
+}
+
+// Add moves the specified file into the instance's temporary directory.
+// If the instance's temporary directory doesn't exist yet, Add will create and lock it.
+// Note: 'path' must be on the same filesystem as the TempDir for os.Rename to work.
+// The caller MUST ensure .Cleanup() is called.
+func (td *TempDir) Add(path string) error {
+	if td.tempDirLock == nil {
+		if err := td.initInstanceTempDir(); err != nil {
+			return fmt.Errorf("initializing instance temp dir failed: %w", err)
+		}
+	}
+	td.tempDirLock.AssertLocked()
+
+	fileName := time.Now().Format("20060102-150405") + "-" + filepath.Base(path)
```

Comment:
One second-level timestamps don’t guarantee uniqueness, computers are fast enough.

If this should manage unique names, maybe just increment a (per-instance) counter?